llama-index-integrations/selectors/llama-index-selectors-notdiamond/llama_index/selectors/notdiamond/base.py

```diff
@@ -0,0 +1,178 @@
+import logging
+import os
+from typing import Sequence
+
+from llama_index.core.llms.llm import LLM
+from llama_index.core.schema import QueryBundle
+from llama_index.core.tools.types import ToolMetadata
+from llama_index.core.base.base_selector import SelectorResult
+from llama_index.core.selectors import LLMSingleSelector
+from llama_index.core.selectors.llm_selectors import _build_choices_text
+
+from notdiamond import NotDiamond, LLMConfig, Metric
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.WARNING)
+
+
+class NotDiamondSelectorResult(SelectorResult):
+    """A single selection of a choice provided by Not Diamond."""
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    session_id: str
+    llm: LLMConfig
+
+    @classmethod
+    def from_selector_result(
+        cls, selector_result: SelectorResult, session_id: str, best_llm: LLMConfig
+    ) -> "NotDiamondSelectorResult":
+        return cls(session_id=session_id, llm=best_llm, **selector_result.dict())
+
+
+class NotDiamondSelector(LLMSingleSelector):
+    def __init__(
+        self,
+        client: NotDiamond,
+        metric: Metric = None,
+        timeout: int = 10,
+        api_key: str = None,
+        *args,
+        **kwargs,
+    ):
+        """
+        Initialize a NotDiamondSelector. Users should instantiate and configure a NotDiamond client as needed before
+        creating this selector. The constructor will raise errors re: required client fields.
+        """
+        # Not needed - we will route using our own client based on the query prompt
+        # Add @property for _llm here
+        _encap_selector = LLMSingleSelector.from_defaults()
+        self._llm = None
+        self._prompt = _encap_selector._prompt
+
+        if not getattr(client, "llm_configs", None):
+            raise ValueError(
+                "NotDiamond client must have llm_configs before creating a NotDiamondSelector."
+            )
+
+        if metric and not isinstance(metric, Metric):
+            raise ValueError(f"Invalid metric - needed type Metric but got {metric}")
+        self._metric = metric or Metric("accuracy")
+
+        self._client = client
+        self._llms = [
+            self._llm_config_to_client(llm_config)
+            for llm_config in self._client.llm_configs
+        ]
+        self._timeout = timeout
+        super().__init__(_encap_selector._llm, _encap_selector._prompt, *args, **kwargs)
+
+    def _llm_config_to_client(self, llm_config: LLMConfig | str) -> LLM:
```

Comment:
Instead of using dedicated integrations, could we not use their LiteLLM integration so that we support all the providers rather than just these providers?

Reply:
We don't route to all of the providers supported by LiteLLM (or LlamaIndex), so I'd prefer to curate this list for now.

Reply:
That's true but in any case, our backend would reject a model if we don't support it no?

Reply:
Discussed offline, we'll leave as-is for the moment but move to LiteLLM if users share feedback.